chore (curriculum) add js fundamental review module transcripts (#58802)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/lecture-understanding-core-javascript-fundamentals/6732c6d4dec34c61850a1276.md

# --description--

Watch the video or read the transcript and answer the questions below.

# --transcript--

What is memory management, and how does it work in JavaScript?

Memory management is an essential concept in programming, but it can be a bit confusing for beginners, let's break it down in simple terms.

> When you run a program including JavaScript code in a web browser it needs memory to store all the information its working with, this includes varibles, functions, objects, basically everything your code creates and uses. Memory management is the process of controlling this memory, allocating it when needed, and freeing it up when it's no longer needed. In some programming languages, developers have to manually manage memory. They need to explicitly tell the computer when to allocate memory for new things and when to free up memory that's no longer needed. This can be powerful but also tricky as for getting to free memory can lead to memory leaks.

Reviewer: ```suggestion
When you run a program, including JavaScript code in a web browser, it needs memory to store all the information its working with. This includes varibles, functions, objects, basically everything your code creates and uses. Memory management is the process of controlling this memory, allocating it when needed, and freeing it up when it's no longer needed. In some programming languages, developers have to manually manage memory. They need to explicitly tell the computer when to allocate memory for new things and when to free up memory that's no longer needed. This can be powerful but also tricky as for getting to free memory can lead to memory leaks.
```

> JavaScript however uses automatic memory management, this means that JavaScript more specifically the JavaScript engine in your web browser takes care of memory allocation and deallocation for you. You don't have to explicitly free memory in your code. This automatic process is often called garbage collection.

Reviewer: ```suggestion
JavaScript however uses automatic memory management. This means that JavaScript (more specifically the JavaScript engine in your web browser) takes care of memory allocation and deallocation for you. You don't have to explicitly free memory in your code. This automatic process is often called garbage collection.
```

Here's how it works in simple terms. First allocation happens when you create a variable, objects or functions in your JavaScript code, memory is automatically allocated to store them. Then you use this allocated memory when you work with these variables, objects or functions in your code. 

> The JavaScript engine has clever ways to figure out when something in memory is no longer needed. Generally if there is no way for your program to access or use a piece of data any more it's considered no longer needed. Periodically the garbage collector runs, it finds a memory that's no longer needed and frees it up making it available for future use. This process happens automatically which is great because it means you don't have to worry about managing memory yourself. 

Reviewer: ```suggestion
The JavaScript engine has clever ways to figure out when something in memory is no longer needed. Generally if there is no way for your program to access or use a piece of data any more it's considered no longer needed. Periodically the garbage collector runs, it finds a memory that's no longer needed and frees it up, making it available for future use. This process happens automatically which is great because it means you don't have to worry about managing memory yourself. 
```

> The JavaScript engine has clever ways to figure out when something in memory is no longer needed. Generally if there is no way for your program to access or use a piece of data any more it's considered no longer needed. Periodically the garbage collector runs it, finds a memory that's no longer needed and frees it up making it available for future use. This process happens automatically which is greeat because it means you don't have to worry about managing memory yourself. 

Reviewer 2: ```suggestion
The JavaScript engine has clever ways to figure out when something in memory is no longer needed. Generally if there is no way for your program to access or use a piece of data any more it's considered no longer needed. Periodically the garbage collector runs, it finds a memory that's no longer needed and frees it up making it available for future use. This process happens automatically which is great because it means you don't have to worry about managing memory yourself. 
```

However, it's still important to understand this concept because you can sometimes accidentally keep references to things you don't need anymore preventing the garbage collector from freeing that memory. For example:

```js
function createLargeArray() {
  let largeArray = new Array(1000000);
  return function() {
    console.log(largeArray.length);
  };
}

let printArrayLength = createLargeArray();
printArrayLength();
```

In this code, even after `createLargeArray` finishes running, `largeArray` can't be garbage collected because the returned function still has access to it. This is a closure, and while closures are useful they can sometimes lead to more memory usage than you might expect. You will learn more about closures in future lecture videos.

> As a beginner you don't need to worry too much about the intricacies of memory management, JavaScript's automatic garbage collection takes care of most things for you. However, as you advance in your JavaScript journey, understanding this concepts can help you write more efficient code especially for larger applications or when working with limited resources. 

Reviewer 2: ```suggestion
As a beginner you don't need to worry too much about the intricacies of memory management, JavaScript's automatic garbage collection takes care of most things for you. However, as you advance in your JavaScript journey, understanding these concepts can help you write more efficient code especially for larger applications or when working with limited resources. 
```

> As a beginner you don't need to worry too much about the intricacies of memory management, JavaScript's automatic garbage collection takes care of most things for you. However, as you advance in your JavaScript journey, understanding these concepts can help you write more efficient code especially for larger applications or when working with limited resources. 

Reviewer: ```suggestion
As a beginner, you don't need to worry too much about the intricacies of memory management. JavaScript's automatic garbage collection takes care of most things for you. However, as you advance in your JavaScript journey, understanding these concepts can help you write more efficient code, especially for larger applications or when working with limited resources. 
```

> Remember, good coding practice like avoiding global variables when possible and being mindfull of what your functions are closing over can help the JavaScript engine manage memory more efficiently.

Reviewer: ```suggestion
Remember, good coding practices, like avoiding global variables when possible and being mindful of what your functions are closing over, can help the JavaScript engine manage memory more efficiently.
```
